drivers: watchdog: stm32: Add the Early Wakeup support for IWDG

juickar opened the pull request:
This PR adds the early wakeup fonctionnality for the IWDG.

juickar committed:
dts: arm: st: Add the interrupt line for the IWDG

Add the interrupt line for the socs with the IWDG implemented.

Signed-off-by: Julien Racki <julien.racki-ext@st.com>

juickar committed:
drivers: watchdog: Add Kconfig option for the IWDG Early Wakeup

Add a Kconfig option that is enabled if the soc's
IWDG support early wakeup.

Signed-off-by: Julien Racki <julien.racki-ext@st.com>

juickar committed:
drivers: watchdog: Add the support of the IWDG Early Wakeup

Add the support of the IWDG early wakeup.
Gives the possibility for the user to set a
callback that will be called before an IWDG reset.

Signed-off-by: Julien Racki <julien.racki-ext@st.com>

mathieuchopstm commented on drivers/watchdog/Kconfig.stm32:
Consider enclosing all options inside a `if IWDG_STM32` instead of adding `depends on` to each option.

mathieuchopstm commented on drivers/watchdog/Kconfig.stm32:
```diff
@@ -30,6 +30,19 @@ config IWDG_STM32_INITIAL_TIMEOUT
+		|| SOC_SERIES_STM32U0X \
+		|| SOC_SERIES_STM32U3X \
+		|| SOC_SERIES_STM32U5X \
+		|| SOC_SERIES_STM32WBAX)
```
I would exclude unsupported series instead since all new ones should have the feature (or introduce a new compatible)

mathieuchopstm commented on drivers/watchdog/wdt_iwdg_stm32.c:
```diff
@@ -174,7 +188,20 @@ static int iwdg_stm32_install_timeout(const struct device *dev,
+		}
+		/* EWI time should be less than reload time but should be at least 2 */
+		ewi_time = (reload > 2U) ? (reload - 1U) : 2U;
+	}
```
1. Create a Kconfig option for the early wake-up threshold instead of hardcoding it
2. Simplify the logic + rewording
```suggestion
        if (reload == 1U) { //i.e., < 2U
            /* Early wake-up is not possible if counter is
             * reloaded with value 1 as the system will be
             * reset immediately when the counter decrements.
             * (Note: reload=1 is NOT RECOMMENDED per RefMan)
             */
             return -EINVAL;       
        }

        /*
         * If Kconfig is higher than reload, set early wake-up
         * as high as possible (= reload - 1); otherwise, set
         * to the requested value.
         */
	uint32_t ewi_time = MIN(CONFIG_EWI_TIME, reload);

	if (ewi_time >= reload) {
		if (reload <= 2U) {
			/* Reload value is too small and EWI time cannot be set */
			return -EINVAL;
		}
		/* EWI time should be less than reload time but should be at least 2 */
		ewi_time = (reload > 2U) ? (reload - 1U) : 2U;
	}
```

Note: as far as I understand, it is valid to have `RLR = EWIT` (since early wake-up occurs when counter is *strictly less than* `EWIT`)?

Note 2: Kconfig should be `range 2 4095`

juickar commented on drivers/watchdog/Kconfig.stm32:
```diff
@@ -30,6 +31,22 @@ config IWDG_STM32_INITIAL_TIMEOUT
+	default 100
+	range 2 4095
+	help
+	  Set the early wake-up interrupt value for IWDG in ticks.
```
done

juickar commented on drivers/watchdog/wdt_iwdg_stm32.h:
```diff
@@ -28,6 +28,14 @@ struct iwdg_stm32_config {
+
+#define IWDG_STM32_DATA(dev) ((struct iwdg_stm32_data *const)(dev)->data)
+
+#define IWDG_STM32_STRUCT(dev) ((IWDG_TypeDef *)(IWDG_STM32_CFG(dev))->instance)
```
Done

juickar commented on drivers/watchdog/wdt_iwdg_stm32.c:
```diff
@@ -134,7 +188,7 @@ static int iwdg_stm32_setup(const struct device *dev, uint8_t options)
 	/* Reload counter just before leaving */
 	LL_IWDG_ReloadCounter(cfg->instance);
-
+	LL_IWDG_DisableWriteAccess(cfg->instance);
```
Done

erwango commented on drivers/watchdog/wdt_iwdg_stm32.c:
```diff
@@ -150,12 +204,17 @@ static int iwdg_stm32_install_timeout(const struct device *dev,
+		data->callback = config->callback;
+#else
 		return -ENOTSUP;
+#endif /* CONFIG_IWDG_STM32_EARLY_WAKEUP */
```
We can probably use `IS_ENABLED()` here

erwango commented on drivers/watchdog/wdt_iwdg_stm32.h:
```diff
@@ -28,6 +28,10 @@ struct iwdg_stm32_config {
+	wdt_callback_t callback;
 };
 
+#define IWDG_STM32_STRUCT(dev) (((const struct iwdg_stm32_config *)dev->config)->instance)
```
Those (`WWDG_STM32_STRUCT` and friends ) should actually be removed rather than added.
They were all over but were cleaned up, but wwdg felt through the cracks, as it seems.

erwango commented on drivers/watchdog/Kconfig.stm32:
```diff
@@ -30,6 +31,25 @@ config IWDG_STM32_INITIAL_TIMEOUT
 
+config IWDG_STM32_EARLY_WAKEUP
+	bool "Early Wakeup Interrupt support for STM32 IWDG"
+	default y
```
As most users didn't need it until now, keep default disabled unless there is a good reason to do otherwise.
Drawback is that we need to enable at least once in wwdg tests using a board overlay to be sure it is compiled in CI.

erwango replied:
Yeah, but has footprint impact.

erwango reviewed (changes requested):
Otherwise LGTM

GeorgeCGV commented on drivers/watchdog/wdt_iwdg_stm32.c:
```diff
@@ -150,12 +204,17 @@ static int iwdg_stm32_install_timeout(const struct device *dev,
 				      const struct wdt_timeout_cfg *config)
 {
 	struct iwdg_stm32_data *data = dev->data;
+	__maybe_unused const struct iwdg_stm32_config *cfg = dev->config;
```
Doesn't appear to be used.

juickar committed:
samples: watchdog: Add CI build test for STM32 IWDG EWI

This commit ensures the CI will run test build  for the STM32
that supports the IWDG early wake-up.

Signed-off-by: Julien Racki <julien.racki-ext@st.com>

mathieuchopstm commented on drivers/watchdog/Kconfig.stm32:
```diff
@@ -30,6 +31,25 @@ config IWDG_STM32_INITIAL_TIMEOUT
 
+config IWDG_STM32_EARLY_WAKEUP
+	bool "Early Wakeup Interrupt support for STM32 IWDG"
+	default n
```
`n` is the default
```suggestion
```
